### PR TITLE
Refactor async service to use an async reconciler interface and handle getting existing resources

### LINKED ISSUE #1871: Refactor Get existing logic into CreateResource() async helper
Currently, we're implementing the same logic to get the existing resource for each service in the client CreateOrUpdateAsync function that has been refactored by #1702. We should try to extract that logic to CreateResource() so it can be shared by all the services, have better unit testing. That will also ensure the clients have minimal business logic and stay as close to wrappers of the SDK as possible.

### PR DESCRIPTION
**Pull request by Jont828**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**: This PR refactors the async service into an implementation of an asynchronous reconciler interface. It also refactors the logic for fetching the existing resource into `CreateResource`, simplifying `CreateOrUpdateAsync`, which is meant to wrap the Azure client calls and not do a lot of heavy computation. The unit tests for the existing async services were also refactored and simplified using a mock of the async reconciler interface.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1871 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Add async reconciler interface and refactor async service to get existing resources
```

### CODE REVIEW

**Comment by Jont828** [comment]
@CecileRobertMichon Here's an idea I was experimenting with for refactoring the logic for getting an existing resource into a helper. Do you think this is in the right direction?

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:79]: no need to check for long running operations here, if we reach this phase it means there is already no running operation

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:72]: does this need to be exported? If it's just a helper for CreateResource, it should start with a lower case

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:71]: comment copy paste

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:97]: let's only print the success log if the error was nil

File: azure/services/vnetpeerings/client.go
Comment (CecileRobertMichon) [inline azure/services/vnetpeerings/client.go:78]: I wonder if we could go even further and move this logic up in async.go too, then `CreateOrUpdateAsync` can become ` CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{})`

File: azure/services/vnetpeerings/client.go
Comment (CecileRobertMichon) [inline azure/services/vnetpeerings/client.go:68]: 👌

File: azure/services/vnetpeerings/client.go
Comment (Jont828) [reply azure/services/vnetpeerings/client.go:78]: Would we be able to do the type check for params, i.e. `natRule, ok := params.(network.InboundNatRule)` if we move this part into the interface?

Comment (CecileRobertMichon) [reply azure/services/vnetpeerings/client.go:78]: no that part would stay in here but you could do the nil check in there, ie.:

async.go:

```
if existing, err := client.Get(ctx, spec); err != nil && !azure.ResourceNotFound(err) {
		return nil, errors.Wrapf(err, "failed to get existing resource %s/%s (service: %s)", rgName, resourceName, serviceName)
	} else if err == nil {
		existingResource = existing
	}

params, err := spec.Parameters(existingResource)
if err != nil {
		return nil, nil, errors.Wrapf(err, "failed to get desired parameters for resource %s/%s (service: %s)", rgName, resourceName, serviceName)
} else if params == nil {
   // nothing to do, return existing, don't call CreateOrUpdateAsync
}
```

and in here:

```
func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter, parameters interface{}) (interface{}, azureautorest.FutureAPI, error) {
...
peering, ok := parameters.(network.VirtualNetworkPeering)
if !ok {
    return nil, nil, errors.Errorf("%T is not a network.VirtualNetworkPeering", params)
}

future, err := ac.peerings.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.OwnerResourceName(), spec.ResourceName(), peering, network.SyncRemoteAddressSpaceTrue)

...
```

**Comment by Jont828** [comment]
@CecileRobertMichon Just pushed some more changes. If that looks good, I can refactor all the current async services on main to use this logic and update the tests for them as well.

File: azure/services/vnetpeerings/client.go
Comment (CecileRobertMichon) [inline azure/services/vnetpeerings/client.go:78]: beautiful 👌

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:113]: IMO the previous comment still makes sense, we're still creating the resource. Getting the existing resource to check if anything needs to be updated before doing a create is an implementation detail

maybe something like
```suggestion
	// No long-running operation is active, so get or create the resource.
```
makes sense?

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:72]: this is one of those situations where it might not make sense to have a helper function. This function is very specific to the CreateResource function above, takes all the same function parameters as input, and returns everything as output, which makes the code a bit awkward. It might make more sense if we were only returning the parameters as output, but here we also need the existing resource to use as result if parameters is nil. Consider moving the logic directly into `CreateResource`, or finding another way to breakout the code into smaller funcs if that makes the function too big / hard to read.

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:122]: this isn't actually an error?

Comment (Jont828) [reply azure/services/async/async.go:122]: I don't think we'd ever reach that block if everything is working correctly. If there are no parameters, then we'd have an existing resource that's up to date, and if we have parameters, then we'd be able to create or update using that. If both `parameters` and `existingResource` are `nil`, then it seems like we'd have violated the specs of the function in our code. Does that sound right to you, or is there a better way to handle that?

Comment (CecileRobertMichon) [reply azure/services/async/async.go:122]: Right, there are really just 3 cases:

if err != nil {
     // something went wrong, return an error
} else if parameters == nil {
    // nothing to do, return existing resource 
}

// otherwise, keep going and create the resource

**Comment by CecileRobertMichon** [comment]
@Jont828 once the PR is ready for review and tests are passing, please squash commits to reflect only distinct logical changes and update the PR title/description to remove WIP and reflect the current state of the PR (ie. the other changes that were made besides get existing logic)

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:103]: > If it does and matches the spec, we can return early.

this is not always true, it is up to the `Parameters()` func to decide.

```suggestion
	// Get the resource if it already exists, and use it to construct the desired resource parameters.
```

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:123]:
```suggestion
	// Create or Update the resource with the desired parameters.
```

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:118]:
```suggestion
		// Nothing to do, don't create or update the resource and return the existing resource.
```

**Review by CecileRobertMichon — APPROVED** [review]
/lgtm
/assign @shysank @mboersma @devigned 

This will make it a lot easier to switch further services to async + unit test them, and ensure we always try to get the existing resource before creating it.

File: azure/services/groups/groups_test.go
Comment (mboersma) [inline azure/services/groups/groups_test.go:49]: I assume `errCtxExceeded` and `fakeFuture` are unused?

File: azure/services/async/interfaces.go
Comment (mboersma) [inline azure/services/async/interfaces.go:45]: It's arguably odd to have `Get` in a `Creator` interface, but I'm ok with it here.

**Review by mboersma — APPROVED** [review]
/lgtm

This seems like a consistent refactoring. Nice work!

**Comment by Jont828** [comment]
Sounds good, thanks for the reviews! I'm going to merge the async disks PR and rebase this before merging.

**Comment by CecileRobertMichon** [comment]
/hold

for rebase after #1711 merges

File: azure/services/async/async.go
Comment (CecileRobertMichon) [inline azure/services/async/async.go:32]:
```suggestion
// Service is an implementation of the Reconciler interface. It handles asynchronous creation and deletion of resources.
```

since it was renamed

File: azure/services/disks/client.go
Comment (CecileRobertMichon) [inline azure/services/disks/client.go:32]: remove the comment?

File: azure/services/disks/client.go
Comment (CecileRobertMichon) [inline azure/services/disks/client.go:39]:
```suggestion
// newClient creates a new disk Client from subscription ID.
 ```

not your PR but while you're here...

File: azure/services/disks/mock_disks/client_mock.go
Comment (CecileRobertMichon) [inline azure/services/disks/mock_disks/client_mock.go:17]: what's going on here?

File: azure/services/disks/mock_disks/doc.go
Comment (CecileRobertMichon) [inline azure/services/disks/mock_disks/doc.go:19]: probably need to remove this line too

File: azure/services/disks/mock_disks/client_mock.go
Comment (CecileRobertMichon) [reply azure/services/disks/mock_disks/client_mock.go:17]: this file should be deleted

File: azure/services/groups/groups.go
Comment (CecileRobertMichon) [inline azure/services/groups/groups.go:115]: you might be able to replace this with `s.Reconciler.Get` or `s.Reconciler.Creator.Get` (I think first one should work) and get rid of the client too

File: azure/services/disks/client.go
Comment (CecileRobertMichon) [reply azure/services/disks/client.go:32]: this is still unresolved

**Comment by CecileRobertMichon** [comment]
@devigned @mboersma PTAL when you get a chance. Hoping to merge this one soon so we can rebase and unblock a bunch of the open async services PRs

File: azure/services/virtualmachines/client.go
Comment (CecileRobertMichon) [inline azure/services/virtualmachines/client.go:37]: do we still need the Client interface here?

**Comment by CecileRobertMichon** [comment]
@Jont828 #1812 merged so you need to rebase this one, the conflicts should be too complex, just make sure you use the prefered logging mechanism from #1812 everywhere

File: test/e2e/data/infrastructure-azure/v1alpha4/cluster-template-kcp-adoption.yaml
Comment (CecileRobertMichon) [inline test/e2e/data/infrastructure-azure/v1alpha4/cluster-template-kcp-adoption.yaml:1]: these files shouldn't be part of the PR, were they removed from .gitignore?

File: azure/services/async/async.go
Comment (CecileRobertMichon) [reply azure/services/async/async.go:33]: looks like the rename was reverted

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm
/assign @devigned

**Review by devigned — APPROVED** [review]
The refactor is pretty clean. However, I'm really looking forward to Go 1.18 and generics to clean up the `interface{}` usage.

Solid work, @Jont828.

**Pull request merged**
